log-warper-core/src/Log/Core/Class.hs

```diff
@@ -0,0 +1,201 @@
+{-# LANGUAGE DefaultSignatures          #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE FunctionalDependencies     #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs               #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+-- | Monad type class for things that have logging ability.
+
+module Log.Core.Class
+       ( -- * Interface
+         MonadLogger (..)
+       , logR
+       , withR
+       , unliftLogAction
+
+         -- * Basic interface implementation
+       , ActionT (..)
+       , usingActionT
+       ) where
+
+import Control.Monad (void)
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Trans.Class (MonadTrans (..))
+import Control.Monad.Trans.Except (ExceptT, mapExceptT, runExceptT)
+import Control.Monad.Trans.Identity (IdentityT (..), mapIdentityT)
+import Control.Monad.Trans.Maybe (MaybeT (..), mapMaybeT)
+import Control.Monad.Trans.Reader (ReaderT (..), ask, mapReaderT)
+import Control.Monad.Trans.State.Strict (StateT (..), get, mapStateT)
+import Data.Kind (Type)
+
+import Log.Core.Action (LogAction (..), cmap)
+
+{- | Type class for monads that have logging ability. In order to log things
+monad @m@ should be able to construct 'LogAction' for some logging record @r@.
+
+__Laws:__
+
+Some laws duplicate @lens@ laws. And maybe of the laws are redundant and can be
+proven from other laws.
+
+* __@Idempotency:@__ @askLogAction@ shouldn't mutate monadic context:
+
+@
+'askLogAction' >> 'askLogAction' ≡ 'askLogAction'
+@
+
+* __@Identity:@__ 'id' function shouldn't change anything:
+
+@
+'withLogAction' id ≡ id
+@
+
+* __@Composition:@__ composition of two 'withLogAction' is a single
+  'withLogAction' with function composition:
+
+@
+'withLogAction' f . 'withLogAction' g ≡ 'withLogAction' (f . g)
+@
+
+* __@PutGet:@__ 'askLogAction' after 'withLogAction' should return modified action.
+
+@
+'withLogAction' f 'askLogAction' ≡ 'fmap' f 'askLogAction'
+@
+-}
+class Monad m => MonadLogger (m :: Type -> Type) (r :: Type) | m -> r where
```

Comment:
I didn't think about this much. But `MonadLogger m r` is indeed equivalent to `MonadReader (LogAction m r) m`. Unfortunately, `MonadReader` type class has functional dependency. This means that in some big constraint alias you can't have:

```haskell
type WorkMode =
    ( MonanReader (LogAction m r) m
    , MonadReader SomethingElse m
    )
```

This won't compile. Which makes `MonadReader` approach less useful.

But what we can actually have is type class like `HasLogAction`. Though for now I'm not sure how it will look like...

Reply:
I am not sure this class should be part of `core`. There are many different approaches to monadic effects and using a type class for a particular action is only one possible option. There are many other options, e. g. the simplest one is explicit passing of `LogAction` to functions which need it. Another option is `HasLogAction` as you said.
Some people don't need this class, but I suppose `core` is intended to contain definitions which (almost) everybody needs in order to use `log-warper`.